File: 백준/Silver/22871. 징검다리 건너기 （large）/징검다리 건너기 （large）.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    int N;
    long K;
    int[] stones;
    boolean[] vis;
    private void solution() {
        input();

        long start = 1;
        long end = (N - 1) * (1L + Math.abs(stones[N-1] - stones[1]));
        while (start <= end) {
            K = (start + end)/2;
            vis = new boolean[N];
            dfs(0);

            if (vis[N-1]) {
                end = K - 1;
            } else {
                start = K + 1;
            }
        }

        System.out.println(start);
    }

    void dfs(int idx) {
        vis[idx] = true;

        for (int i = idx + 1; i < N; i++) {
            long energy = (i - idx) * (1L + Math.abs(stones[i] - stones[idx]));

            if (energy <= K && !vis[i]) {
                dfs(i);
            }
        }
    }

    void input() {
        try {
            var br = new BufferedReader(new InputStreamReader(System.in));
            N = Integer.parseInt(br.readLine());
            stones = new int[N];
            var st = new StringTokenizer(br.readLine());
            for (int i = 0; i < N; i++) {
                stones[i] = Integer.parseInt(st.nextToken());
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void main(String[] args) throws IOException {
        new Main().solution();
    }
}

File: 백준/Silver/22871. 징검다리 건너기 （large）/징검다리 건너기 （large）.cc
#include <iostream>
#include <vector>

using namespace std;

#define MAX 1'000'000'000'000'000LL;
int stones[5000];

int main() {
	int n = 0;
	cin >> n;

	for (int i = 0; i < n; i++)
	{
		cin >> stones[i];
	}

	long long memo[5000];
	memo[n-1] = 0;

	for (int i = n-2; i >= 0; i--)
	{
		memo[i] = MAX;
		
		for (int j = i + 1; j < n; j++)
		{
			long long energy = (j - i) * (1LL + abs(stones[i] - stones[j]));
			memo[i] = min(memo[i], max(energy, memo[j]));
		}
	}

	cout << memo[0] << '\n';
}